feat(25.04): libc6-dev slices

lczyk opened the pull request:
# Proposed changes
A bunch of other PRs (see below) depend on this, and it's a bit tricky slice. This PR cherrypicks the `libc6-dev` SDF in order to make other PRs smaller.

## Related issues/PRs

This PR is a member of a chain of PRs which depend on one another. The first commit is a squashed version of everything below it. Please review the 'root' PRs first.

#633 - ROCKS-2270/libc6-dev-for-25.04 **(this PR)**
└ #644 - ROCKS-2270/binutils-for-25.04
....└ #648 - ROCKS-2270/cpp-for-25.04
........└ #649 - ROCKS-2270/gcc-14-for-25.04
............└ #610 - ROCKS-2271/rust-1.84-25.04
................└ #634 - ROCKS-2271/cargo-1.84-25.04

Cherrypicked from #610, #598

### Forward porting

#664

## Checklist

* [x] I have read the [contributing guidelines](
https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md)
* [x] I have tested my changes ([see how](https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md#7-test-your-slices-before-opening-a-pr))
* [x] I have already submitted the [CLA form](
https://ubuntu.com/legal/contributors/agreement)

lczyk commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,120 @@
+      /usr/lib/*-linux-*/libresolv.a:
+      /usr/lib/*-linux-*/librt.a:
+      /usr/lib/*-linux-*/libutil.a:
+        arch: [ amd64, arm64, ppc64el, riscv64, s390x]
```
lol. cheers copilot. :laughing:

lczyk commented:
added `priority` since its related to a lot of active PRs and would help a lot with pushing them forward

zhijie-yang commented on slices/libcrypt-dev.yaml:
```diff
@@ -0,0 +1,32 @@
+
+    # All of these are symlinks to so's in the libcrypt1 package
+    contents:
+      /usr/lib/*-linux-*/libcrypt.so:
```
I believe you can just put it together with `libcrypt.a` in the same slice, and put `libcrypt1_libs` in the essential list.

zhijie-yang commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,120 @@
+      /usr/lib/*-linux-*/crtn.o:
+      /usr/lib/*-linux-*/gcrt1.o:
+      /usr/lib/*-linux-*/grcrt1.o:
+        arch: [ amd64, arm64, riscv64, s390x ]
```
```suggestion
        arch: [amd64, arm64, riscv64, s390x]
```

Let's make the formatting consistent across different files. We don't put extra spaces inside the brackets.

zhijie-yang reviewed (changes requested):
Thanks for the PR. Here are some points we can talk about.

lczyk commented on slices/libcrypt-dev.yaml:
```diff
@@ -0,0 +1,32 @@
+    contents:
+      /usr/include/crypt.h:
+
+  core-libs:
+    contents:
+      /usr/lib/*-linux-*/libcrypt.a:
```
sure to both this and the other comment on `libcrypt-dev`. i've basically just modeled this quickly on how `libc6-dev` looks, but have no problems with putting the two together here

cjdcordeiro replied:
same comment as above, and highlighting @zhijie-yang comment too about the possibly unnecessary split

clay-lake commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,135 @@
+      # Note: libc.so here is an ld script that tells the linker to look up the
+      # functions from libc_nonshared.a if they are not found in the libc.so.6
+      # from libc6.
+      /usr/lib/*-linux-*/libc.so:
```
Should we include `ld` if we are including a linker script? I think this depends on the use case, and perhaps the user wants to use a different linker.

clay-lake commented on slices/libcrypt-dev.yaml:
```diff
@@ -0,0 +1,32 @@
+package: libcrypt-dev
```
Should we include the package config files? 
- /usr/lib/x86_64-linux-gnu/pkgconfig/libxcrypt.pc
- /usr/lib/x86_64-linux-gnu/pkgconfig/libcrypt.pc

lczyk replied:
ok. added the pklgconfig slice, but not as essential of libs

clay-lake reviewed (approved):
Few minor questions below but otherwise LGTM

cjdcordeiro commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,135 @@
+    contents:
+      /usr/share/doc/libc6-dev/copyright:
+
+  core-libs:
```
We need some further explanation to what this is. During a functional split, `minimum`, `core`, etc., have a meaning, as defined in the chisel-docs. `core-libs` is a new thing. I'm not saying it's wrong, but it begs the question about its purpose and relation to `libs` (if there's even one)

lczyk replied:
talked with @zhijie-yang and proposed a new version. will mark this comment as resolved and re-request review from @clay-lake too

cjdcordeiro reviewed (changes requested):
let's have @zhijie-yang also re-review this

alesancor1 commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,153 @@
+      /usr/include/*-linux-*/gnu/**.h:
+      /usr/include/*-linux-*/ieee754.h:
+      /usr/include/*-linux-*/sys/**.h:
+      /usr/include/a*.h:
+      /usr/include/arpa/*.h:
+      /usr/include/byteswap.h:
```
Why are we using the glob with `a*.h`, `d*.h`, `e*.h` but directly referencing other headers like `byteswap.h`, `complex.h`, `cpio.h`, etc.? Why not just include all .h files under /include? `/usr/include/*.h`

alesancor1 replied:
Right that's a good point. However, being this the case I'd rather list all the files individually because:
* Every time someone creates a slice that puts something under `/usr/include` a conflict will happen:
   * This slice would need to be modified as part of the PR
   * Even if the new slice is not depending on this `libc6-dev` ones (meaning no conflict happens on CI), the conflicting files are still there.
* If at some point we wanted to refactor and glob this files instead, it is safer as globing will add files rather than remove them, avoiding potential regressions.

Note that this is more a limitation with chisel on how conflicts are handled, so an issue could be raised there (if not yet raised).

lczyk replied:
i've pushed the full unglob to test it in CI while i look for some deeper paths (deeper than `/usr/include/` for which it would make sense to glob. i will push that separately so if we decide to go back to full unglob, it's just a couple of reverted commits away

lczyk commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,153 @@
+essential:
+  - libc6-dev_copyright
+
+slices:
```
^

lczyk commented:
based on the overall consensus i've renamed the `posix-libs` slice to `core` and i've added it as essential for `libutil`, `librt` etc. ( good spot @cjdcordeiro ) some of the tests for the behavior there will need to come in #644 since we don't have the linker yet, and using the host linker (from outside of the chroot) for test would be tricky and error-prone imo. will work on that now and add these to #644, as well as updating it to use the new slice names.

`core` slice represents the libraries one will always need, since they are the basic things you need to create a binary. the tests should check that + check that, for example, if we install the `libm` slice but then manually delete `core`, we cannot compile a working binary -- aka test asserting / showing that `core` libs are indeed core.

lczyk commented:
based on `dpkg --search` on a full installation of ubuntu 25.04 (`find -maxdepth 1 -type f -printf "$PWD/%P\n" | xargs -n1 dpkg --search`) i found that it seems to be safe to glob `bits` and `src`. i've globbed these, but left arch-specific globs explicitly specified. i also left all shallower things unglobbed.

its possible that `/usr/include/net` or `/usr/include/netinet` could be globbed too, but the savings there are mich smaller.

lczyk commented:
just so it does not get lost in a response to the review comment, [here](https://github.com/lczyk/libc6-dev-deb-data) is the data dump of `dpkg -c` of `libc6-dev` .h files in all supported archs.

alesancor1 reviewed (approved):
This is amazing work, seems you got to understand libc6 quite in depth, which is great! (some bits are a bit over explained in comments I'd say, but not against it)

Apart from what others said I have no further comments, LGTM! Let's get those nits fixed quick and we good to go on my end :+1:

lczyk commented on slices/libc6-dev.yaml:
```diff
@@ -0,0 +1,659 @@
+    essential:
+      - libcrypt-dev_headers
+      - linux-libc-dev_linux-headers
+    contents:
```
good catch on that btw. thanks heaps :)

zhijie-yang reviewed (approved):
Looks good to me now! Many thanks!

clay-lake reviewed (approved):
Nice work 💪
